File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/GroupPlugin.java
package mymonitorhub.agent.core;

import com.fasterxml.jackson.databind.JsonNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

/**
 * A plugin that hosts multiple child plugins and runs them on their
 * individual hibernate schedules from a single thread.
 * Mirrors C# GroupPlugin.
 */
public class GroupPlugin extends AbstractPlugin {

    private static final Logger log = LoggerFactory.getLogger(GroupPlugin.class);
    private final List<AbstractPlugin> pluginList = new ArrayList<AbstractPlugin>();

    @Override
    public void initialize(JsonNode groupPluginElement, String threadTitle) {
        super.initialize(groupPluginElement, threadTitle);

        JsonNode pluginNodeList = groupPluginElement != null
                ? groupPluginElement.get("plugin") : null;
        if (pluginNodeList == null || !pluginNodeList.isArray()) return;

        for (int i = 0; i < pluginNodeList.size(); i++) {
            JsonNode pluginElement = pluginNodeList.get(i);
            String className = pluginElement.path("class-name").asText(null);
            if (className == null) continue;

            try {
                Class<?> clazz = Class.forName(className);
                Object instance = clazz.newInstance();
                if (!(instance instanceof AbstractPlugin)) {
                    throw new RuntimeException("class (" + className + ") must implement AbstractPlugin");
                }
                AbstractPlugin plugin = (AbstractPlugin) instance;
                plugin.initialize(pluginElement, getTitle());
                pluginList.add(plugin);
            } catch (Exception e) {
                log.error("Failed to load group plugin class {}: {}", className, e.getMessage());
            }
        }
    }

    public void displayPlugins() {
        for (AbstractPlugin plugin : pluginList) {
            log.info("    Plugin: {} - runs every {} seconds",
                    plugin.getTitle(), plugin.getHibernateSeconds());
        }
    }

    @Override
    protected void execute() {
        for (AbstractPlugin plugin : pluginList) {
            if (plugin.canRun()) {
                plugin.start();
            }
        }
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/GroupPlugin.cs
using System;
using System.Collections;
using System.IO;
using System.Reflection;
using Newtonsoft.Json.Linq;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    public class GroupPlugin : AbstractPlugin
    {
        private static readonly ILogger Logger = Log.ForContext<GroupPlugin>();
        private readonly ArrayList _pluginList = new ArrayList();

        public override void Initialize(JToken groupPluginElement, string threadTitle)
        {
            base.Initialize(groupPluginElement, threadTitle);
            var pluginNodeList = groupPluginElement["plugin"] as JArray;

            if (pluginNodeList != null)
            {
                var length = pluginNodeList.Count;
                for (var index = 0; index < length; index++)
                {
                    var pluginElement = pluginNodeList[index];

                    var className = pluginElement["class-name"]?.ToString();
                    var assemblyName = pluginElement["assembly"]?.ToString();

                    var assembly = assemblyName == null ? GetType().Assembly : LoadPluginAssembly(assemblyName);

                    // create an instance of the class from the assembly
                    var o = assembly.CreateInstance(className!);

                    if (!(o is AbstractPlugin))
                    {
                        throw new Exception($"class ({className})must implement AbstractPlugin");
                    }
                    var plugin = (AbstractPlugin)o;
                    plugin.Initialize(pluginElement, Title ?? string.Empty);

                    _pluginList.Add(plugin);
                }
            }
        }

        /// <summary>
        /// Resolves a plugin assembly path relative to the install directory and validates
        /// it does not escape via path traversal. Prevents CWD-hijacking (A-H5).
        /// </summary>
        private static Assembly LoadPluginAssembly(string assemblyName)
        {
            var baseDir = AppContext.BaseDirectory;
            var fullPath = Path.GetFullPath(Path.Combine(baseDir, assemblyName));
            if (!fullPath.StartsWith(baseDir, StringComparison.OrdinalIgnoreCase))
                throw new InvalidOperationException(
                    $"Plugin assembly path '{assemblyName}' resolves outside the install directory ('{baseDir}'); load rejected.");
            return Assembly.LoadFrom(fullPath);
        }

        public virtual void DisplayPlugins()
        {
            foreach (AbstractPlugin plugin in _pluginList)
            {
                Logger.Information("    Plugin: " + plugin.Title + " - runs every " + plugin.HibernateSeconds +
                                  " seconds");
            }
        }

        protected override void Execute()
        {
            // this thread runs every 30 seconds as defined in GroupPluginThread.hibernateSeconds however each plugin runs as
            // every 2 minutes or longer

            foreach (AbstractPlugin plugin in _pluginList)
            {
                if (plugin.CanRun())
                {
                    plugin.Start();
                }
            }
        }
    }
}